### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by ratamacue**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. 

You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_number.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
+  end
+
+  def number
+    @number
+  end
+
+  def to_i
+    n = @number.gsub!(/[^0-9]/, "")
+    n.to_i
```
Comment (StrainedQuacking) [inline phone_number.rb:31]: This `gsub!` will affect `@number`, removing any formatting we wanted on it. This `to_i` function could be more elegantly done in one line with the following:

```
@number.gsub(/[^0-9]/, "").to_i
```

Without the bang operator on `gsub`, we will only return the result without affecting `@number`, and we can just call `to_i` to do all of our operation on one line.

File: phone_number.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+  end
+
+  def call(message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(@number, message)
+  end
+
+  def text_and_call(message)
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
```
Comment (StrainedQuacking) [inline phone_number.rb:22]: Looks like we're passing `message` here twice (once for our number value, once for our message). This should just be `(@number, message)`

File: phone_number.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def find_area_code
+    open_paren = @number.index('(')
+    close_paren = @number.index(')')
+    if open_paren
+      if close_paren
+        @number[(open_paren + 1)..(close_paren - 1)]
+      end
+    else
```
Comment (StrainedQuacking) [inline phone_number.rb:47]: We don't have an `else` case for the `if close_paren`. If we are given a number with an open parentheses but no close, we may have some unintended consquences (Most likely, the `else` will just include our open parentheses in the area code, which we do not want)

File: phone_number.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      @number.gsub!(/[^0-9]/, "")
+      @number[0..2]
+    end
+  end
+
+  def find_exchange
+    @number.gsub!(/[^0-9]/, "")
```
Comment (StrainedQuacking) [inline phone_number.rb:54]: We're calling `.gsub!` here and in the other utility functions for `format_number`. Furthermore we have an extra case in `find_area_code` just for a number that isn't formatted, which then will just call `gsub!` if we did not find parentheses. I'd recommend instead just calling `.gsub!` before line 37 in `format_number`. We'll cut down on extra code, extra unnecessary writes to the variable, and we won't need the parentheses flow control in `find_area_code`.

File: phone_number.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    @number[3..5]
+  end
+
+  def find_line_number
+    @number.gsub!(/[^0-9]/, "")
+    @number[-4..@number.length]
```
Comment (StrainedQuacking) [inline phone_number.rb:60]: This negative indexing works for our purposes, but for readability, I would recommend following indexing similar to the previous two functions: `[6..9]`

File: phone_number_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  describe "#to_i" do
+    it "is the number as an int" do
+      new_number = PhoneNumber.new("555-867-5309")
+      expect(new_number.to_i).to eq(5558675309)
+    end
+  end
+
+  describe "#text" do
+    it "calls PhoneActions#text!" do
+      expect_any_instance_of(PhoneActions).to receive(:text!)
```
Comment (StrainedQuacking) [inline phone_number_spec.rb:21]: We should probably ensure in our tests that here and on line 28 as well we are passing the correct arguments to `PhoneActions`.

File: phone_number_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  describe "#number" do
+    it "is a formatted number" do
+      expect(phone_number.number).to eq("(555) 867-5309")
+    end
+  end
+  
+  describe "#to_i" do
+    it "is the number as an int" do
+      new_number = PhoneNumber.new("555-867-5309")
```
Comment (StrainedQuacking) [inline phone_number_spec.rb:14]: If we're going to set up a new `PhoneNumber` with hyphens to test `#to_i` here, we should also probably test that the hyphenated number gets formatted correctly as well. Better yet, we can set up test cases with multiple different edge-case inputs that use an array to test that `format_number` and `to_i` work with each one.

File: phone_number.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
+require_relative "./phone_actions"
+
+class PhoneNumber
+  USERNAME = ENV['ACTIONS_USERNAME']
+  PASSWORD = ENV['ACTIONS_PASSWORD']
+
+  def initialize(number)
+    @number = number
+    @number = format_number
```
Comment (StrainedQuacking) [inline phone_number.rb:9]: I'm not a huge fan of storing the number as formatted here. It's hard to tell in what cases we want the number to have formatting and where we may want it to just be an `int`. I would almost say we should just call our `gsub!` here and save our number as our regex adjusted string, and then use `format_number` in anywhere that we want to pass the number with formatting. This way, we can know that our number will always be in the format of "XXXXXXX" when accessed as `@number`, and as "(XXX) XXX-XXXX" when calling `format_number`, and finally just an int of XXXXXXX when calling `to_i`.

File: phone_number.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  def format_number
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
+  end
+
+  def find_area_code
+    open_paren = @number.index('(')
+    close_paren = @number.index(')')
+    if open_paren
+      if close_paren
+        @number[(open_paren + 1)..(close_paren - 1)]
```
Comment (StrainedQuacking) [inline phone_number.rb:45]: If we get passed input with some unusual parentheses placement [i.e. "12)2483(1234"], this current flow will have unexpected results, and we may not get the numbers we're looking for.
